pkg/core/notary_test.go

```diff
@@ -0,0 +1,668 @@
+			}
+			// then the whole request should be removed, i.e. there are no completed transactions
+			finalizeWithError = false
+			require.NoError(t, bc.AddBlock(bc.newBlock()))
+			checkSigTx(t, requests, len(requests), false)
+			checkFallbackTxs(t, requests, false)
+
+			// it's OK to remove unexisting fallback transaction
+			ntr1.OnFallbackRemoval(requests[0])
+			require.NoError(t, bc.AddBlock(bc.newBlock()))
+			checkSigTx(t, requests, len(requests), false)
+			checkFallbackTxs(t, requests, false)
+		})
+
+		t.Run("Multisignature request", func(t *testing.T) {
+			nSigs, nKeys := 3, 5
+			// check OnNewRequest with finalization error
+			finalizeWithError = true
+			requests := checkCompleteMultisigRequest(t, nSigs, nKeys, false)
+			checkMultisigTx(t, nSigs, requests, len(requests), false)
+			checkFallbackTxs(t, requests, false)
+
+			// make fallbacks valid and remove one fallback
+			_, err := bc.genBlocks(int(nvbDiffFallback))
+			require.NoError(t, err)
+			unlucky := requests[nSigs-1]
+			ntr1.OnFallbackRemoval(unlucky)
+			// then m out of n fallbacks should be completed
+			finalizeWithError = false
+			require.NoError(t, bc.AddBlock(bc.newBlock()))
+			checkMultisigTx(t, nSigs, requests, len(requests), false)
+			checkFallbackTxs(t, requests[:nSigs-1], true)
+			require.Nil(t, completedTxes[unlucky.FallbackTransaction.Hash()])
+			// but the rest (n-m) out of n fallbacks shouldn't be completed because main tx has been collected by the moment they were sent
+			checkFallbackTxs(t, requests[nSigs:], false)
+
+			// make another request
+			finalizeWithError = true
+			requests = checkCompleteMultisigRequest(t, nSigs, nKeys, false)
+
+			// make fallbacks valid and then remove all of them
+			_, err = bc.genBlocks(int(nvbDiffFallback))
+			require.NoError(t, err)
+			for i := range requests {
+				ntr1.OnFallbackRemoval(requests[i])
+			}
+			// then the whole request should be removed, i.e. there are no completed transactions
+			finalizeWithError = false
+			require.NoError(t, bc.AddBlock(bc.newBlock()))
+			checkMultisigTx(t, nSigs, requests, len(requests), false)
+			checkFallbackTxs(t, requests, false)
+
+			// it's OK to remove unexisting fallback transaction, i.e. there still shouldn't be any completed transactions after this
+			ntr1.OnFallbackRemoval(requests[0])
+			require.NoError(t, bc.AddBlock(bc.newBlock()))
+			checkMultisigTx(t, nSigs, requests, len(requests), false)
+			checkFallbackTxs(t, requests, false)
+		})
+	})
+}
+
+func TestNotary_UpdateNotaryNodes(t *testing.T) {
+	bc := newTestChain(t)
+	defer bc.Close()
+	acc1, ntr1 := getTestNotary(t, bc, "./testdata/notary1.json", "one")
+	randomKey, err := keys.NewPrivateKey()
+	require.NoError(t, err)
+	// set account for the first time
+	ntr1.UpdateNotaryNodes(keys.PublicKeys{acc1.PrivateKey().PublicKey()})
+
+	t.Run("account is already set", func(t *testing.T) {
+		ntr1.UpdateNotaryNodes(keys.PublicKeys{acc1.PrivateKey().PublicKey(), randomKey.PublicKey()})
```

Comment:
What does this test check?

Reply:
It checks the following part of code:
```
        if n.currAccount != nil {
		for _, node := range notaryNodes {
			if node.Equal(n.currAccount.PrivateKey().PublicKey()) {
				return
			}
		}
	}
```
I.e. when Notary nodes are updated, and the new list includes the old account. In this case, we don't have to change our Notary account.

I need to move this test to `services` pkg and after that I'll add `require` check. So I'll move it soon.